### PR TITLE
Move conversion.Scheme to runtime

### PR DESCRIPTION
**Pull request by smarterclayton**
There is only one type registry to rule them all, and it is runtime.Scheme.

Moved tests to runtime, preserved most of the tests that didn't depend on scheme dealing
with interface{}.

Continuation from #17922

### CODE REVIEW

**Comment by smarterclayton** [comment]
@kubernetes/sig-api-machinery

**Comment by smarterclayton** [comment]
@k8s-bot test this

File: pkg/runtime/scheme.go [hunk 1]
```diff
@@ -28,46 +28,101 @@ import (
 // Scheme defines methods for serializing and deserializing API objects. It
 // is an adaptation of conversion's Scheme for our API objects.
```
Comment (lavalamp) [inline pkg/runtime/scheme.go]: Fix comment?

File: pkg/runtime/scheme.go [hunk 2]
```diff
@@ -75,111 +130,218 @@ func NewScheme() *Scheme {
...
+// or an error if it's not a pointer or is unregistered.
 func (s *Scheme) ObjectKinds(obj Object) ([]unversioned.GroupVersionKind, error) {
-	return s.raw.ObjectKinds(obj)
+	v, err := conversion.EnforcePtr(obj)
+	if err != nil {
+		return nil, err
+	}
+	t := v.Type()
+
+	gvks, ok := s.typeToGVK[t]
```
Comment (lavalamp) [inline pkg/runtime/scheme.go]: Should there be a lock on this map? Do we disallow non-init time registrations?

File: pkg/runtime/scheme.go [hunk 3]
```diff
@@ -75,111 +130,218 @@ func NewScheme() *Scheme {
...
+	t := v.Type()
+
+	if _, ok := s.typeToGVK[t]; !ok {
+		return false, false
+	}
+	_, ok := s.unversionedTypes[t]
+	return ok, true
 }
 
 // New returns a new API object of the given version ("" for internal
```
Comment (lavalamp) [inline pkg/runtime/scheme.go]: "" is no longer internal, right?

File: pkg/runtime/scheme.go [hunk 4]
```diff
@@ -75,111 +130,218 @@ func NewScheme() *Scheme {
...
-	s.raw.Log(l)
+	s.converter.Debug = l
 }
 
-// AddConversionFuncs adds a function to the list of conversion functions. The given
-// function should know how to convert between two API objects. We deduce how to call
-// it from the types of its two parameters; see the comment for
-// Converter.RegisterConversionFunction.
+// AddIgnoredConversionType declares a particular conversion that should be ignored - during conversion
+// this method is not invoked.
```
Comment (lavalamp) [inline pkg/runtime/scheme.go]: Can you clarify this? What does ignored mean? I think the comment from the version of this function that you're deleting is clearer.

File: pkg/runtime/scheme.go [hunk 5]
```diff
@@ -75,111 +130,218 @@ func NewScheme() *Scheme {
...
+			return err
+		}
+	}
+	return nil
 }
 
 // AddDeepCopyFuncs adds a function to the list of deep-copy functions.
 // For the expected format of deep-copy function, see the comment for
 // Copier.RegisterDeepCopyFunction.
 func (s *Scheme) AddDeepCopyFuncs(deepCopyFuncs ...interface{}) error {
```
Comment (lavalamp) [inline pkg/runtime/scheme.go]: Maybe in the future we can take this out of scheme. It seems totally unrelated to me.

**Comment by lavalamp** [comment]
LGTM with just a few minor changes

File: pkg/runtime/scheme.go (hunk 2 above)
Comment (smarterclayton) [inline pkg/runtime/scheme.go]: We don't allow it, expect it, or do it, but we don't disallow it.

File: pkg/runtime/scheme.go (hunk 3 above)
Comment (smarterclayton) [inline pkg/runtime/scheme.go]: Correct

File: pkg/runtime/scheme.go (hunk 4 above)
Comment (smarterclayton) [inline pkg/runtime/scheme.go]: Ok, will update.  Tried to pick the comment that was better for these, but
this one I think was worse.

File: pkg/runtime/scheme.go (hunk 5 above)
Comment (smarterclayton) [inline pkg/runtime/scheme.go]: Agree - honestly this should probably just be a method on each struct.

Comment (smarterclayton) [inline pkg/runtime/scheme.go]: Will update those.

**Comment by smarterclayton** [comment]
Updated comments, applying LGTM

**Pull request merged**